### ISSUE TITLE
New permission override is not applied on new course

### ISSUE THREAD

**Issue by dionysius** [frankkoch/moodle-mod_studentquiz#270]
I just merged #260 yesterday and awaited test results. During final manual test the permission override is not applied in this exact order:

- create new course
- create studentquiz activity
- enrol student to course
- login as student
- view studentquiz activity

I get the following error: `Sorry, but you do not currently have permissions to do that (access question edit tab questions).`

This means, that the context override is either not executed or the enrolment event did not see the user/the role of that user.

The exact order matters: Enroling another user after fixes the problem or creating the studentquiz activity after enrolment does not produce that error either - so in those cases the context override is applied.

Tested on Moodle 3.9 and 4.0dev. SQ on current master (https://github.com/frankkoch/moodle-mod_studentquiz/commit/609b2097aea3799435f763955154ba78b04ee5da)

**Comment by dionysius** [comment]
I tracked down the issue to https://github.com/frankkoch/moodle-mod_studentquiz/blob/609b2097aea3799435f763955154ba78b04ee5da/classes/access/context_override.php#L98, as I assumed:
The function `get_roles_used_in_context`:
1. When adding a user with a new role - returns without the one beeing added.
2. When removing an enrolment, returns without the one beeing removed.
3. When changing an enrolment role (using the edit button on the role), nothing happens [1], it only reads the correct roles when changing the enrolment itself (using the gear icon)

The way it looks the behaviour of 1. and 2. *is inconsistent*. The should *either both write* already their result to the database *or both shouldn't*.

For 3. we are missing an event:

[1] Nothing happens: Event `\core\event\user_enrolment_updated` *is** not fired!*
![image](https://user-images.githubusercontent.com/1341084/95100067-02669000-0731-11eb-905f-6c799108dea8.png)

[2] The update event is sent, but the "problem" already happened, as you can't change the role in that dialogue.
![image](https://user-images.githubusercontent.com/1341084/95100193-22964f00-0731-11eb-8c1f-cac64be00459.png)

**Comment by dionysius** [comment]
It is even worse, I can't make a workaround since the select role is **not part of the event**:
```
core\event\user_enrolment_created Object
(
    [data:protected] => Array
        (
            [eventname] => \core\event\user_enrolment_created
            [component] => core
            [action] => created
            [target] => user_enrolment
            [objecttable] => user_enrolments
            [objectid] => 15
            [crud] => c
            [edulevel] => 0
            [contextid] => 49
            [contextlevel] => 50
            [contextinstanceid] => 5
            [userid] => 2
            [courseid] => 5
            [relateduserid] => 3
            [anonymous] => 0
            [other] => Array
                (
                    [enrol] => manual
                )

            [timecreated] => 1601913093
        )

    [logextra:protected] => 
    [context:protected] => context_course Object
        (
            [_id:protected] => 49
            [_contextlevel:protected] => 50
            [_instanceid:protected] => 5
            [_path:protected] => /1/3/49
            [_depth:protected] => 3
            [_locked:protected] => 0
        )

    [triggered:core\event\base:private] => 1
    [dispatched:core\event\base:private] => 
    [restored:core\event\base:private] => 
    [recordsnapshots:core\event\base:private] => Array
        (
        )

)
```

**Comment by timhunt** [comment]
Instead of using enrol events, would it be better to look at role_assigned/role_unassigned events? That will also work for roles assigned in other context beside course context.

**Comment by timhunt** [comment]
I think that, consistently, role_assigned/role_unassigned events fire after the DB has been updated. (They are fired in accesslib.php)

**Comment by dionysius** [comment]
Thanks for the tip! I definitely should have to look into them

**Comment by dionysius** [comment]
This is excellent, role_(un)assigned event is available also in Moodle 3.5+ and they catch the "change" of [1] by sending assigned and then the unassigned event (for us the order does not matter).

This will make the code a lot easier and I can remove moodle version "if"-hacks as well.

**Comment by dionysius** [comment]
Let the tests begin: https://github.com/frankkoch/moodle-mod_studentquiz/pull/271
